archunit/src/main/java/com/tngtech/archunit/core/domain/JavaClass.java

```diff
@@ -1407,9 +1480,41 @@ Interfaces withGenericTypes(Set<JavaType> genericTypes) {
         }
     }
 
+    private static class EnclosingDeclaration {
+        static final EnclosingDeclaration ABSENT = new EnclosingDeclaration(Optional.<JavaCodeUnit>absent(), Optional.<JavaClass>absent());
+
+        private final Optional<JavaCodeUnit> enclosingCodeUnit;
+        private final Optional<JavaClass> enclosingClass;
+
+        private EnclosingDeclaration(Optional<JavaCodeUnit> enclosingCodeUnit, Optional<JavaClass> enclosingClass) {
+            this.enclosingCodeUnit = checkNotNull(enclosingCodeUnit);
+            this.enclosingClass = checkNotNull(enclosingClass);
+        }
+
+        boolean isPresent() {
```

Comment:
Well, but logically there can't be an enclosing code unit, if there is no enclosing class, right? The method I wanted is really "is the enclosing declaration present", so I would not rename the method. You could argue to check both optionals then, I just thought enclosing code unit present, but enclosing class not is an invalid state :man_shrugging: (which could also be validated in the constructor to make it clear :thinking:)

Reply:
I should take another look (which I can't right now, as I'm travelling), but does it really make sense to have an `EnclosingDeclaration` with two empty `Optional`s? I wonder whether it'd be nicer to model this as `Optional<EnclosingDeclaration>`, where `EnclosingDeclaration` always has an enclosing class. Of course you _could_ do runtime checks in the constructor, but I guess it would be better if this was already handled by the compiler.

Reply:
I just thought this way we can encapsulate more into this type :man_shrugging: I tried it the other way and it just makes everything longer and more convoluted as you need more case distinctions. And `Superclass` already follows the same pattern, I think it's okay to encode the optional part into a type (although I see your point, but I think it's a small internal detail, so for me it's still alright and maybe not worth to put a lot more effort into :man_shrugging:).